File: src/main/java/com/micronaut/bug/api/EnumParam.java
package com.micronaut.bug.api;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;
import io.micronaut.serde.annotation.Serdeable;
import lombok.Getter;
import lombok.RequiredArgsConstructor;

import java.util.Arrays;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

@Serdeable
@Getter(onMethod = @__(@JsonValue))
@RequiredArgsConstructor
public enum EnumParam {

    VALUE_1(10),
    VALUE_2(20),
    ;

    public static final Map<Integer, EnumParam> VALUE_MAPPING = Map.copyOf(Arrays.stream(values())
        .collect(Collectors.toMap(v -> v.value, Function.identity())));

    private final int value;

    @JsonCreator
    public static EnumParam fromValue(int value) {
        return VALUE_MAPPING.get(value);
    }
}

File: src/main/kotlin/com/micronaut/bug/controller/EnumParam.kt
package com.micronaut.bug.controller

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonValue
import io.micronaut.core.annotation.Introspected

@Introspected
enum class EnumParam(
    @get:JsonValue
    val value: Int
) {

    VALUE_1(10),
    VALUE_2(20),
    ;

    companion object {

        @JvmField
        val VALUE_MAPPING = entries.associateBy { it.value }

        @JsonCreator
        @JvmStatic
        fun fromValue(value: Int?): EnumParam {
            return VALUE_MAPPING[value]!!
        }
    }
}
